### DISCUSSION TITLE
Adding multi-core support for YT-Spammer-Purge

### DISCUSSION THREAD

**Discussion by justinktn** [ThioJoe/YT-Spammer-Purge#585] [Ideas]
Is it possible if we can add multi-core support for much faster scanning of spam comments?

**Comment by Masamune3210** [comment]
Not even slightly limited by hardware lol, it's a python script. It's likely more a api limitation

**Reply by justinktn** [reply]
I think it is limited by hardware because when I try it on a laptop that has a low-power CPU, it is not so fast. But when I try it on a desktop (Intel i7 CPU) which has high clock speed, it is much much faster. When I checked on Task Manager, I can see that it is only using one of my cores. I think maybe it is possible that we can add support for multi-core for reporting spam comments faster.

**Reply by Masamune3210** [reply]
I promise you a python script like this is not slowed down by your hardware unless you have other things stealing the CPU or you just have really slow hardware lol

**Comment by Firecul** [comment]
IIRC he tried multi-threading for one feature but it wasn't as fast at the time so there may have been another bottleneck.

**Reply by ThioJoe** [reply]
Yea I'll try and take a crack at it again at some point probably. It's not just that it wasn't as fast, it was WAY slower. There's a branch where you can try it out. It should work, but you'll see what I mean. I'm not really experienced enough to know what the problem is.

There's no doubt that the program could benefit greatly if I could get multiprocessing working in some places. Especially checking for duplicates and checking sets of comments against the filters. It could mean being able to add more complex filters because it wouldn't slow things down nearly as much.

Multithreading could help with checking against the filter while waiting for the next API response, but from my understanding, in python multithreading still only allows doing one thing at a time. Where multi processing actually allows running multiple actions simultaneously on different cores, but is a bit more complicated.

There's also async, but unfortunately it wouldn't help at all because the python API modules for the Google API don't support async, and aren't even thread safe.

**Reply by TechStudent10** [reply]
> Yea I'll try and take a crack at it again at some point probably. It's not just that it wasn't as fast, it was WAY slower. There's a branch where you can try it out. It should work, but you'll see what I mean. I'm not really experienced enough to know what the problem is.
> 
> There's no doubt that the program could benefit greatly if I could get multiprocessing working in some places. Especially checking for duplicates and checking sets of comments against the filters. It could mean being able to add more complex filters because it wouldn't slow things down nearly as much.
> 
> Multithreading could help with checking against the filter while waiting for the next API response, but from my understanding, in python multithreading still only allows doing one thing at a time. Where multi processing actually allows running multiple actions simultaneously on different cores, but is a bit more complicated.
> 
> There's also async, but unfortunately it wouldn't help at all because the python API modules for the Google API don't support async, and aren't even thread safe.

There **HAS** to be an `async` version of the Google API client.

**Reply by rachmadaniHaryono** [reply]
quick google search give me this https://pypi.org/project/aiogoogle/
